### ISSUE TITLE
Run JavaScript in rendered HTML

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor.js#6]
If the generated HTML contains `<script>` tags, they need to be handled explicitly when embedding the HTML into another document. Provide an API that performs this function.

In fact, it might be a good idea to provide an API that handles embedding the result into an HTML element. We can use Opal's native interface to interact with the DOM.

**Comment by Mogztter** [comment]
It's a different approach but it can be related to https://github.com/asciidoctor/asciidoctor/issues/2539 ?

**Comment by mojavelinux** [comment]
I'm not sure this is the same thing. It's more about providing a bridge when using Asciidoctor.js for tooling. When a standalone document is created, the browser executes the whole document in the normal way. However, when an embedded document is created, it may need to be inserted into a DOM at runtime. This is where an API would come in handy that can insert the embeddable document, then ensure that any scripts in that document get executed just like they would when the page loads.

Now, this may not be an Asciidoctor.js concern directly. Perhaps it's some sort of document loader for browser-based tooling.

**Comment by Mogztter** [comment]
> Now, this may not be an Asciidoctor.js concern directly. Perhaps it's some sort of document loader for browser-based tooling.

I think it's an integration issue. The same thing is true when you insert an HTML text into a DOM at runtime.
If you want `<script>` elements inserted with `.innerHTML` to be executed then there's a few ways to do it. In the browser extension, I'm using something like:

```js
const html = "<h1>Hello</h1><script>console.log('hello')<\/script>"
const contentElement = document.getElementById('content')
contentElement.innerHTML = html
const scripts = contentElement.querySelectorAll(':scope > script')
for (let script of scripts) {
  var scriptElement = document.createElement('script')
  scriptElement.innerText = script.innerText
  document.body.appendChild(scriptElement)
}
```

But depending on your context (and security concerns) you may use another solution.
We could add a documentation page on the user manual to make it clear that `<script>` won't be executed in this use case ? What do you think ?

**Comment by Mogztter** [comment]
> I'm not sure this is the same thing. 

I was suggesting that it's related to asciidoctor/asciidoctor#2539 because if we have an API to register inline scripts then the embedded document won't contain `<script>` elements anymore. Since we could ask the processor to give us the list of inline `<script>` we could do:

**pseudo-code.js**
```js
const html = asciidoctor.convert(input, options)
const inlineScripts = asciidoctor.getInlineScripts() // get the inline scripts
var scriptElement = document.createElement('script')
scriptElement.innerText = inlineScripts.map(script => script.getText()).join('\n')
document.body.appendChild(scriptElement)
```

**Comment by mojavelinux** [comment]
> We could add a documentation page on the user manual to make it clear that <script> won't be executed in this use case ?

A documentation page would be a good first step.

> Since we could ask the processor to give us the list of inline <script> we could do:

This feels like we are coupling the processor too much with the converter and output format. I could see this as perhaps a method on the converter.

**Comment by mojavelinux** [comment]
So perhaps a method on the converter? Or even a method contributed by the converter (only available when that converter is used)?

**Comment by stken2050** [comment]
As I have asked earlier in gitter, it should be nice to have the converter method to obtain DOM other than html strings.

I think to obtain just list of Script DOM with working-context is rather complicated, but instead, how about to provide `convertToDOM` method?

This would be simpler and more versatile in use cases,. Developers simply append DOM to the document, then script should work, and if want to obtain/remove/modify the Script dom, they can do via standard DOM  APIs.

**Comment by mojavelinux** [comment]
> convertToDOM

I'd really like to avoid having converter-specific methods show up on the main API. That really goes against the design of Asciidoctor, IMO. While I understand the DOM is very important for a use case like yours, it gives special treatment to a specific output format when AsciiDoc is supposed to be backend agnostic.

What it could be, however, is a separate converter, like html-dom that returns an object instead of a string. That would very much be consistent with the design of Asciidoctor.

Another option is to have the converter put scripts into the catalog. The catalog is designed for this purpose. (Then you can return a document object and access them via doc.catalog.scripts). The catalog is a repository of references that were found during processing. Although there are built-in keys, that doesn't mean there couldn't be others.

**Comment by stken2050** [comment]
I'm not certain if I understand your explanation 100%, and I would like to mention in general design not limited to my cases.

1. As far as I observe, conversion to HTML document is a primary feature of Asciidoctor/.js.

2. For the HTML document, the DOM/render for the browser is the only usage I can think of.

3. "Run JavaScript in rendered HTML" is in fact, the topic for that, and as @Mogztter has illustrated in the code, and like you saying: "to interact with the DOM." I think what we discuss here is the `convertToDOM`.

>to provide an API that handles embedding the result into an HTML element. 

Isn't this an API `convertToDOM`?

Sorry, I'm confused, and please let me know what I misunderstand. Thanks.

**Comment by mojavelinux** [comment]
> conversion to HTML document is a primary feature of Asciidoctor/.js.

Conversion to HTML is built-in, yes. Conversion to HTML is popular, yes. But the primary feature of Asciidoctor is that it is backend/output agnostic. We cannot allow backend-specific methods to leak into the top-level API.

> DOM/render for the browser is the only usage I can think of.

For your use case, it may be the only usage. I assure you there are plenty of other use cases to which we have to cater.

> "Run JavaScript in rendered HTML" is in fact, the topic for that, and as @Mogztter has illustrated in the code, and like you saying: "to interact with the DOM."

We agreed in that other issue that this is, in fact, and intermediary utility that applies for tools that inject HTML into an existing frame. In short, it's an integration issue. So it's not a responsibility of the core API.

**Comment by mojavelinux** [comment]
To build on that last point, when HTML is injected into an existing frame, by default, scripts don't execute by default. So this is a concern that has to do with transferring the result that Asciidoctor produces into a "live" environment. I would expect that to be a feature of the library that does the injecting.

**Comment by stken2050** [comment]
Thanks. In fact, I just thought to provide API to obtain just for script elements is not versatile.

In terms of a "live" environment, let me mention. 

To be honest, I won't use this script live reflection. If the live load is only once, the best solution would be just saving HTML as a file and load to the broswer, and for the live mutable environment, executing the scripts for the whole block everytime does have a big performance issue. 

They are the heaviest process. Just imagine in the HTML document, you have 10+ code blocks to be highlighted, 10+ latex for mathjax/katex, the whole process from scratch by injected script every time are far from efficiency.

So, here's my solution I actually implemented. 

For stem, if a script is injected, stem attribute is true, and for highlightjs, role/class is added as so.

After the confirmation, the non-working scripts are removed, CSS remains.

https://github.com/stken2050/asciidoc-live-electron-viewer/blob/master/src/render_build.ts#L75-L81

As I asked in the chat, I know there are only two script groups to be injected: hightlight.js and katex is used depending on the attribute. Doing so in node/electron environment, I can control the behavior of the script execution for each code/stem blocks.

For performance, I implement a cashing system using dictionary pattern.

https://github.com/stken2050/asciidoc-live-electron-viewer/blob/master/src/render_script.ts#L39-L52

So, in theory, the rendering process by hightlight/katex for an identical code never repeated.
